### PR TITLE
Add sparse compressed fake tensor support

### PR DESCRIPTION
**Pull request by pearu**
As in the title.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120920

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @jcaip @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

### CODE REVIEW

File: torch/_subclasses/meta_utils.py [hunk 1]
```diff
@@ -321,6 +325,64 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
                         with torch.enable_grad():
                             r = r.clone()
                             r._coalesced_(t.is_coalesced())
+                elif is_sparse_compressed(t):
+                    is_leaf = safe_is_leaf(t)
+
+                    def mk_meta():
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py:331]: This function is weirdly a lot more complicated than the one in sparse_coo above, not sure if it's needed or not yet

Comment (ezyang) [reply torch/_subclasses/meta_utils.py:331]: I guess the problem is you don't have a constructor that just directly takes sparse/dense dim and allocates those tensor under the hood, do you...

**Review by ezyang — APPROVED** [review]

File: torch/_subclasses/meta_utils.py (hunk 1 above)
Comment (pearu) [reply torch/_subclasses/meta_utils.py:331]: Exactly, we don't have `_sparse_coo_tensor_with_dims`-alike for sparse compressed tensors. I could expose one as an aten operator but does it matter if it is either implemented in Python (as above) or in C++? IIRC, their specification will be the same: construct as sparse compressed tensor with indices and values as empty tensors with device='meta' and `nnz==0`.
What do you think?

> ... not sure if it's needed or not yet

It surely appears to be needed as without it a bunch of dynamo tests are failing.

**Comment by thiagocrepaldi** [comment]
@pearu would https://github.com/pytorch/pytorch/issues/123345 be a corner case for this pr?

**Pull request closed without merging**